server/src/main/java/io/crate/expression/symbol/SymbolVisitors.java

```diff
@@ -95,17 +97,21 @@ public static Iterable<AnalyzedRelation> extractAnalyzedRelations(@Nullable Symb
     @SuppressWarnings({"SuspiciousMethodCalls", "unchecked"})
     private static class IntersectionVisitor<T> extends SymbolVisitor<Void, Void> {
 
-        private final Collection<T> haystack;
+        private final Map<T, T> haystack;
         private final Consumer<T> consumer;
 
-        public IntersectionVisitor(Collection<T> haystack, Consumer<T> consumer) {
-            this.haystack = haystack;
+        private IntersectionVisitor(Collection<T> values, Consumer<T> consumer) {
+            this.haystack = new HashMap<T, T>(values.size());
+            for (T t : values) {
+                haystack.put(t, t);
+            }
```

Comment:
I wonder if we can avoid Map creation and keep using Collection.

We can introduce a static helper method, iterating over Collection and retuning haystack item on equality.